### DISCUSSION TITLE
Repeat systems

### DISCUSSION THREAD

**Discussion by jim-ec** [bevyengine/bevy#8614] [Q&A] [answered]
I want to repeat a system based on a count variable. How would I achieve that?

A non-compiling snippet outlining my problem:

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        // Does not work, add_systems only accepts tuples:
        .add_systems(std::iter::repeat(my_system).take(5))
        .run();
}

fn my_system() {
    println!("Hello world");
}
```

**Comment by TimJentzsch** [comment, marked as answer]
This depends on multiple factors:

- Do you want to repeat the system's functionality _within_ a frame or _for a defined number of frames_?
- Is the number of repetitions fixed once the app has been constructed or could it change every frame?
- Which part of your app should be able to modify the number of repetitions?

For example, if you want to repeat the system's functionality within a frame, have the number of repetitions be variable and modified by other systems in your app, I would do the following:

- Introduce a `Resource` struct wrapping `usize` to define the repetition count:
    ```rs
    #[derive(Resource)]
    struct RepetitionCount(usize);
    ```

- Initialize the resource to the count you want to start with.
- Modify your system's function signature to pull in the resource:
    ```rs
    fn my_system(repetition_count: Res<RepetitionCount>, ...) { ... }
    ```
- Wrap the body of your system in a loop to implement the repetition:
    ```rs
    for _ in (0..repetition_count.0) { ... }
    ```
- In the system that should modify the repetition count, pull in `ResMut` of the count and change it
- Schedule the modifying system to run before `my_system`

(I didn't check if the example code actually compiles so you might need to adjust it a bit)

**Reply by jim-ec** [reply]
Thank you. I was hoping there was some kind of system combinator, but this solution also does the job.
